### PR TITLE
Integrate 1d improvements

### PR DESCRIPTION
**Pull request by randommm**
#### Submission Checklist

- [ ] Run unit tests: `./runTests.py test/unit`
- [ ] Run cpplint: `make cpplint` (for some reason is not working on my computer)
- [x] Declare copyright holder and open-source license: see below

#### Summary:
New signature for integrate_1d, see http://discourse.mc-stan.org/t/adding-1d-integrator-to-stan/3222

#### Intended Effect:
Allow.

#### How to Verify:
New tests.

#### Side Effects:
User must specify dummy x_r and x_i even if it's not necessary for the function being integrated.

#### Documentation:
Updated.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):

Marco Inacio

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by syclik** [comment]
@randommm, the doc isn't building properly. Please fix the doxygen errors.

**Comment by randommm** [comment]
@syclik @bob-carpenter tests passed, ready for review

**Comment by bob-carpenter** [comment]
Thanks much.  I can probably get to it tomorrow.  Then I can start in on the grammar side, which should be relatively easy.

**Comment by bob-carpenter** [comment]
I'm taking a look.  I'm going to pull this down and make some edits myself---it'll be easier than requesting them through a review.  Then someone else should probably review it.

**Comment by bob-carpenter** [comment]
Comments on just one file:  `stan/math/rev/mat/functor/integrate_1d_tsc.hpp`

I checked this out and started editing to clarify the doc, but it's going to need *much* clearer doc.  I can't figure out what the arguments to `integrate_1d_tsc()` are supposed to be, nor am I clear on what the gradients are used for, nor whether their memory is being handled soundly.  

The doc at the very top needs to be clearer on template instantiations.  I'm pretty sure that `Tx_r` has to be `vector<double>` or `double`, but that's not stated.  It's not enough to doc `T_x_i` as "Type of x_i"---you need to say what the legal instantiations are, at least somewhere in the doc if not directly in the tparam doc.

Sorry to come back now with these deeper questions, but I'm trying to spend more time diving into the semantics of the pull requests and less time on superficial formatting issues.  I started trying to rewrite the doc to be more explicit, but I realized I didn't understand what it was supposed to be doing, so I'm going to have to punt this back to @randommm.  It was particularly unclear to me how `T_x_i` is involved as it's not mentioned in the doc, but seems to also have two signatures in the two examples.  So I'm guessing there are 8 instantiations, but I want to make sure they all work---this is what they'd look like in Stan.

```
real integrate_1d(F f, real lb, real ub, real[] params, real[] data_r, int[] data_i, real atol, real rtol);
real integrate_1d(F f, real lb, real ub, real[] params, real[] data_r, int data_i, real atol, real rtol);
real integrate_1d(F f, real lb, real ub, real[] params, real data_r, int[] data_i, real atol, real rtol;
real integrate_1d(F f, real lb, real ub, real[] params, real data_r, int data_i, real atol, real rtol);
real integrate_1d(F f, real lb, real ub, real params, real[] data_r, int[] data_i, real atol, real rtol);
real integrate_1d(F f, real lb, real ub, real params, real[] data_r, int data_i, real atol, real rtol);
real integrate_1d(F f, real lb, real ub, real params, real data_r, int[] data_i, real atol, real rtol);
real integrate_1d(F f, real lb, real ub, real params, real data_r, int data_i, real atol, real rtol);
```

I was suggesting we only do the first, as we have done with the other functions.  That would presumably allow things like this `integrate_1d_tscg` function to be massively simplified so we don't have eight possible test configurations.  

@syclik---So this is another question about how much of the combinatorics we need to test.  For this one, I'd say we need to test all of them.

Missing argument doc on `gradient_of_f`.  I take it that's because our doxygen checks aren't strict enough to make sure all the parameters and returns are documented.

Does `.grad()` cleanup memory in that call to `gradient_of_f`?  And what's to stop pollution of adjoint values on the inputs after this (like `param`)---I was expecting something like `value_of` to prevent that?  Nothing on the outside is going to zero all the adjoints before passing down gradients.  So we need to do testing of this in a larger framework where we make sure our final derivatives of the integral are right on the outside.  

@syclik If you understand this better than me, feel free to take over.  I'm going to be diving into MPI next and can come back to this after that, but that's unlikely to be for another week or so, and this one looks like it's going to need a lot of attention.

(Until I can figure out how to view my own reviews so I can verify change requests, I'm going to start just putting them in comments rather than inline.)

**Comment by bgoodri** [comment]
What is the status of this? Would it be easier to just use Boost's functions?
http://discourse.mc-stan.org/t/1d-numerical-integration-again/3767?u=bgoodri
They don't have a relative tolerance argument, only an absolute one (the difference between consecutive levels), and you can't integrate the partial derivatives simultaneously, but other than that it seems to be good and easy to wrap.

**Comment by bgoodri** [comment]
In that case, I am thinking that the signature of the user's function needs to be
```
real foo(real x, real xc, real [] theta, real [] x_r, int [] x_i);
```
and the signature of the function in Stan needs to be either
```
integrate_1d(functor f, real a, real b); 
```
or
```
integrate_1d(functor f, real a, real b, 
                     real tolerance, real error, real L1, int levels); 
```
where the extra arguments on the second line default to the default values in Boost. However, it internally needs to call different Boost functions depending on the values of `a` and `b`:

1. `a` and `b` both finite -> call tanh_sinh
2. `a` and `b` both infinite -> call sinh_sinh
3.  otherwise call exp_sinh with `a` and `b`

Do we want to allow the user to specify `max_halvings`?
https://www.boost.org/doc/libs/1_66_0/libs/math/doc/html/math_toolkit/double_exponential/de_levels.html

**Comment by randommm** [comment]
@bob-carpenter sorry, I just saw the comment you made on Feb 23, until now I didn't knew there was something else to do on this (though the language side was all left that was left).

The implementation I made support a lot of different signatures as you said, when I started working on this I thought that was the right way (there was no "anonymous" vectors and arrays in the language at that time and I didn't checked the ODE solver very well before started) but later I thought it was OK to keep them on the API and just expose some to the language.

If you prefer boost implementation, then maybe the code can be reused to wrap-it-around Stan.

**Comment by bgoodri** [comment]
I haven't looked into it extensively, but my impression is that what Boost
implemented is at least as good as what John Cook put on his blog. Plus,
Boost can work with x and x complement to achieve higher precision and
there are a few different Boost quadrature routines for different
situations that have the same or similar signatures. I think the only
downside is that @randommm did some stuff to integrate the gradient
simultaneously with the user-defined function, whereas with Boost we would
be autodiffing under the integral.

**Comment by betanalpha** [comment]
The analytic gradients are just running the 1D solver over
the autodiff of the integrand, so the approach that Marco
used can be immediately carried over to the boost integration.

I do think that we should have separate functions for
the four possible cases finite and infinite bounds and not
rely on std::inf or equivalent to determine which function to
call.  It will be easier for the user to determine what’s going
on with explicit calls.

**Comment by bob-carpenter** [comment]
> On Apr 5, 2018, at 8:29 PM, Michael Betancourt <notifications@github.com> wrote:
> 
> ...
> I do think that we should have separate functions for
> the four possible cases finite and infinite bounds and not
> rely on std::inf or equivalent to determine which function to
> call. It will be easier for the user to determine what’s going
> on with explicit calls.

I like having clearly designated functions for different situations.

What happens if we allow bonds and someone passes in plus or minus infinity?

**Comment by bgoodri** [comment]
This situation is unlike ODEs where we can't tell whether the thing is
stiff or not and have to let the user choose an algorithm.

What is gained by making the user specify integrate_1d_exp_sinh(f) instead
of integrate_1d(f, 0, positive_infinity())? All you are doing is giving the
user a (considerable) positive probability of choosing the wrong or
suboptimal integration algorithm for their problem.

**Comment by bgoodri** [comment]
I don't think that the fact that we haven't switched on whether something
negative_infinity() or positive_infinity() means that much if we have not
had a reason to before adding quadrature.

But it does not do much good to say integrate_1d_exp_sinh(f) and
integrate_1d_sinh_sinh(f) are "explicit" if half the time the user is
choosing the wrong one and possibly getting invalid results when we know
the preferable algorithm to use just by looking at the limits. It is just
as explicit conceptually to say that integrate_1d(f, 0,
positive_infinity()) represents the area under the curve defined by f
between 0 and positive infinity; it is just less explicit what algorithm is
computing that area. We have log1m(x) rather than
log1m_truncated_taylor_series(x) because the fact that it uses different
Taylor series approximations depending on the value of x is not that
relevant to the user when we can choose the best approximation just by
looking at the value of x.

On Fri, Apr 6, 2018 at 4:22 PM, Michael Betancourt <notifications@github.com
> wrote:

> Adding an entirely new UX element is going to cause
> confusing one way or another. No where else in the
> language do we use negative_infinity() or positive_infinity()
> like this. Adding three functions is just as easy and
> makes it more explicit what exactly is going on.
>
>

**Comment by betanalpha** [comment]
The names shouldn’t refer to the algorithms but the bound
types, i.e. integrate_1d_lower_inf, integrate_1d_upper_inf,
etc.  We had a brief discussion of this way back in the thread.

**Comment by bgoodri** [comment]
OK, I misunderstood you. But you think having

   1. integrate_1d_bounded(f, a, b)
   2. integrate_1d_unbounded(f)
   3. integrate_1d_lower_inf(f, a)
   4. integrate_1d_neginf_upper(f, b)

is better than

integrate_1d(f, a, b)

? I don't see what we are really gaining. If someone calls
integrate_1d_bounded(f, a, b) but b has overflowed to infinity, they are
not in good shape. So, to be safe, they would have to write

if (is_inf(b)) integrate_1d_lower_inf(f, a);

else integrate_1d_bounded(f, a, b);



On Fri, Apr 6, 2018 at 6:03 PM, Michael Betancourt <notifications@github.com
> wrote:

> The names shouldn’t refer to the algorithms but the bound
> types, i.e. integrate_1d_lower_inf, integrate_1d_upper_inf,
> etc. We had a brief discussion of this way back in the thread.
>
>

**Comment by betanalpha** [comment]
Exactly when numbers overflow will be system specific
and so relying on implicit conversion to an internal
infinity specifier in the C++ will lead to unreproducible
outcomes.  I’d rather the user make that decision
explicitly or utilize a function so that they know what
implicit (and possibly also not reproducible) rule is
being used.

**Comment by bob-carpenter** [comment]
Thinking of this in terms of documentation, I like the idea of a single function where we pass in positive or negative infinity.  It's a bit slower that way because it involves dynamic dispatch, but presumably any time here will be massively dominated by the actual integration.

The other place we use this construct in the language is in lower- and upper-bounds for parameters.  If you include a lower bound of negative infinity, it's like not having a bound.

If we do go with multiple versions, I think it's easier to name them by the arguments they provide rather than the ones they leave out.  It's also consistent with how we're trying to name distributions (with descriptions of argument scales suffixed).

```
integrate1d(f);
integrate1d_lb(f, L);
integrate1d_ub(f, U);
integrate1d_lub(f, L, U);
```

But it really does seem easier to just use a single

```
integrate1d(f, L, U);
```

and require both bounds, but allow them to be infinite.

**Comment by betanalpha** [comment]
Yes the bounds with with positive and negative infinity that
way but does anyone actually use them?  Everyone just
drops the lower/upper when the correspond to negative/positive
infinity.

I won’t argue if people want to move forward with that pattern,
we’ll just need extensive documentation about all of the possible
bounds patterns in the manual.

**Comment by bgoodri** [comment]
I wouldn't think we would default the bounds to infinity; the user should
have to specify them. The tolerances I think we could have defaults for
like with ODEs.

On Sun, Apr 8, 2018 at 8:26 PM, Michael Betancourt <notifications@github.com

**Comment by betanalpha** [comment]
Another way of saying it, I’d have no problem if we could
write signatures like

integrate1d<lower=l>(f)
integrate1d<lower=l, upper=positive_infinity()>(f)

Then the pattern would exactly match what we do in the
variable declarations.  

But if we try to mimic that functionality partially using only
negative/positive_infinity() in the bounds then I don’t think 
that it’s close enough that we get any benefit of a common 
UX.  

Anyways, just add a bunch of doc to whatever signature
ends up being used.

**Comment by bob-carpenter** [comment]
I see the appeal, but we don't have any other functions like this and I'm reluctant to introduce new syntax (or overload old syntax) for this one case.  

> integrate1d<lower=l>(f)
> integrate1d<lower=l, upper=positive_infinity()>(f)

**Comment by sakrejda** [comment]
I suppose the syntax would be appropriate for cdf's and the incomplete gamma as well... that would confuse plenty of people but cut down I  the number of functions...

**Comment by bob-carpenter** [comment]
If we could motivate a bunch of bounded functions it might make sense to add it.  I don't think we'd get much confusion about static vs. dynamic evaluation from people importing C++ template intuitions.

**Comment by syclik** [comment]
What's the status of this pull request? @randommm, please include the copyright holder.

Does this change affect the Stan library? (It looks like it does.)

Should the function signature be changed?

**Comment by bbbales2** [comment]
@syclik @randommm I'd be happy to finish this up if that's okay? I've been looking at it recently.

**Comment by bgoodri** [comment]
I think we decided to go with what is in Boost now.

**Comment by randommm** [comment]
Added copyright holder.

I don't know what was the decision on the matter of whether changing the 
signature of this function or not.

@bgoodri, will any of the code here be reused to implement the boost 
version? If so, maybe working from this branch would be good, otherwise, 
maybe then close this PR?

**Comment by bob-carpenter** [comment]
Is it that the current implementation works but we think the Boost version will work better?  If that's the case, maybe we should merge it and worry about improving it later.

Or is there something missing in the current implementation that we need?

**Comment by randommm** [comment]
Isn't it possible to use the wrappers I made to have it work with Boost 
instead Cook's implementation?

**Comment by bgoodri** [comment]
With some modifications to incorporate the extra functionality in Boost,
presumably. The main ones are integration over infinite intervals on one or
both sides and that the function Boost integrates can (and should, to avoid
cancellation) take two arguments (with the second being the distance from
the point to the closest endpoint).

**Comment by syclik** [comment]
I think the question to answer is if the C++ function changes are appropriate. In this pull request there's a new function signature.

**Comment by syclik** [comment]
Here's the summary of the discussion:
- @bgoodri believes we can go with Boost for the implementation
- @bbbales2 has volunteered to implement this summer using Boost

Can someone verify that the Boost implementation is good enough for us to use and replace our current implementation?

@randommm, do you mind if I close this pull request for now?

File: stan/math/rev/mat/functor/integrate_1d_tsc.hpp [hunk 1]
```diff
@@ -22,83 +21,86 @@ namespace stan {
...
  * @param tre target relative error.
  * @param tae target absolute error.
  * @param param additional parameters to be passed to f.
+ * @param x_r additional data to be passed to f.
+ * @param x_i additional integer data to be passed to f.
  * @param msgs stream.
  * @return numeric integral of function f.
  */
-template <typename F, typename G, typename T_param>
+template <typename F, typename G, typename T_param, typename T_x_r,
```
Comment (syclik) [inline stan/math/rev/mat/functor/integrate_1d_tsc.hpp]: We don't need template parameters for `T_x_r` and `T_x_i`. These should be removed and where used, just include the primitive types: `double` and `int`.

**Comment by sakrejda** [comment]
@bbbales2 I think @syclik asked me the get this PR moved along.  I'm happy to do that, I think the first thing that's required is the set of signatures in `stan-dev/math` that this function should support (I read the discussion above and it seems like it's not clear).  If we have a set of signatures we can avoid a lot of confusion and make clearer decisions about what will be supported.  Since the Stan language does not need to use the same signatures lets avoid all the Stan language syntax issues.  Once we have our signatures, we'll ask for comments from the language developers to make sure we don't make their life more difficulty but no need to resolve exactly how they implement things.

@bbbales2 : 1) please let me know if/when you think you'll be able to doc. what signatures we will need; 2) please let me know if you want me to be involved in figuring signatures out too, otherwise I'll assume it's on you.  Sound good?

**Comment by bgoodri** [comment]
If we are going to use Boost, then the signatures should be essentially the
same as for the Boost functions. Like with ODEs, we can have a version
where the user specifies tolerances and a version with fewer arguments that
calls the former with its default tolerances. The main thing in order to
get good accuracy in difficult cases is to pass two unary functions: one
that inputs the thing being integrated over and the other which inputs the
complement of the thing being integrated over. This avoids the cancellation
of a subtraction inside the function. The other thing is maybe we want a
version that passes in a gradient function, so the internal calculations
can be done in doubles, although they have to be done for each scalar
element of the gradient function.

**Comment by bbbales2** [comment]
As far as signatures go, the signature of the outward facing function should be:

`real integrate_1d(F f, real lb, real ub, real[] params, real[] x_r, int[] x_i, real atol, real rtol);`

Boost only has one tolerance, so that'll change, but otherwise it makes sense to just be as consistent with the ODE/algebra solvers as much as possible.

==

For the function signature of the thing to be integrated, here's the Boost page with an example of the cancellation error thing: https://www.boost.org/doc/libs/1_67_0/libs/math/doc/html/math_toolkit/double_exponential/de_tanh_sinh_2_arg.html. I have basically no experience with these types of integrators, but if @bgoodri thinks this is worth watching out for, let's do it. It'll be easy for folks to not use it if they don't want to, and apparently it was important enough for the Boost people to dedicate a page to.

So the function to be integrated should be:
`real f(real x, real xc, real[] theta, real[] x_r, real[] x_i)`

==

For the limits at infinity, let's allow passing in positive_infinity() and negative_infinity() as lower bounds and upper bounds. This doesn't change the current pull req, and is easy to implement. I don't think this would break things even if we moved to allow parameters as integration limits?

==

The thing that makes me nervous is the version of integrate_1d that takes a function as well as a function that computes its gradients. This isn't done anywhere else in Stan, and it'd be easy for someone to mess up their gradients in a lot of ways and get mysterious errors. Naturally the upside is it's way faster and could be used super effectively.

When I was playing with integrals with infinite limits as well, it's easy for autodiff to fail. Boost might ask for a function evaluated at 1e300, which can easily turn into infs, which can easily turn into NaNs (this happens even for `e^(-x^2 / sigma^2)` on [-inf, inf]. But anyway it's already in the math library, so there's nothing to do on it in this pull req, I just thought I'd mention it since I think there will need to be some talkin' about it before this moves to the interfaces.

**Comment by sakrejda** [comment]
So to complete this signature:

```
real integrate_1d(F f, real lb, real ub, real[] params, real[] x_r, int[] x_i, real atol, real rtol);
```

1) I'm guessing based on past discussion that we're only allowing `var` and `double` types, not `fwd` or `fwd<rev>`.  Is this correct?

2) I'm guessing we can have
-  `rev` or `double` for: `lb`, `ub`, `params`
- `double[]` for: `x_r`
- `int[]` for `x_i` 
- `double` for a single combined `tolerance` parameter.
Is this correct?

3) I read the link to Boost's "Handling functions with large features near an endpoint...", in the context of this PR I'm not sure how you're thinking about it: is that an implementation detail or does it result in another signature? Is this something the implementation falls back in if some other method fails?

4) We can always expose another signature that allows users to include gradients later.  First we should get this Boost version in, test it, and provide some examples that help users understand when they should expect it to just work and when it fails.  Adding the feature where you can include hand-crafted gradients will require a lot of additional doc and use cases unless we want the user list traffic to explode.

@bbbales2  If you can answer these I can take a look at the code as it stands (do you have a branch?) and comment on whether I think there are any barriers to moving forward (beyond the implementation, you and Ben obviously have the implementation under control).

**Comment by bgoodri** [comment]
For 3., the signature needs to start with two functions, where the second
one takes the complementary input.

**Comment by bbbales2** [comment]
re 1, yes, let's limit to double and var (like the ODE solver here). That okay @syclik?

re 2, no var for ub or lb. We aren't allowing parameters to be endpoints (yet)

re 3, the signature for the functions passed to integrate_1d change. It's not included in the original pull req here, so I don't think you need to include it now. I'll make this change before the 1d integrator stuff gets to the interfaces though. It's pretty simple (just passing an extra argument around).

**Comment by sakrejda** [comment]
So sounds like we're punting 3. to another PR, punting `var` for lb/ub to another PR, and the only signature outward-facing signature is:

```
real integrate_1d(F f, real lb, real ub, real[] params, real[] x_r, int[] x_i, real tolerance);
```

Where the only `var` types are in `params`.  The bounds are `double` as is `x_r` and `x_i` is an `int` array.  That makes sense and I can see why you're going through with Boost here.  

@bob-carpenter I think this will be straightforward to integrate into the language, any concerns we should address before @bbbales2 brings the code in line with this?

@bgoodri thanks for the background about the longer-term goals for including the different integrator approaches.  Correct me if I'm wrong but it should be fine to do this one at a time since each one is basically adding a feature.

**Comment by bgoodri** [comment]
I think the difference is that I wouldn't call numerical accuracy a
feature. There are examples in the Boost documentation and at
http://discourse.mc-stan.org/t/1d-numerical-integration-again/3767/2?u=bgoodri
where in order to get the right answer, you have to use the two function
approach. I suppose you can say it is like the situation with nonstiff and
stiff ODE solvers, where the stiff solver has the "feature" that yields the
correct solution to stiff systems. But the still solver has a cost that it
takes longer than does the nonstiff solver on nonstiff systems. The
`integrate_1d` function does not fall in that category. The two function
version has essentially the same speed as the one function version in the
case where the one function version is sufficiently accurate. If we expose
a one function version now and the two function version later, the two
function version will be underutilized. Even if we expose them at the same
time, the two function version will be underutilized. The only way to not
underutilize the two function version is to not expose the one function
version, at least not to the Stan language.

On Mon, Jun 4, 2018 at 1:27 PM, Krzysztof Sakrejda <notifications@github.com

**Comment by bgoodri** [comment]
I misremembered things. It is one function with two arguments: `x` and `xc`. And then you branch inside it.

**Comment by bob-carpenter** [comment]
If we use Boost, then we should be able to template out a version to work with forward mode.   I would be OK with a version that doesn't include forward mode.

**Comment by bbbales2** [comment]
@sakrejda this is ready for reviewin'. I think I got everything in place, but lemme know what I screwed up.

**Comment by bgoodri** [comment]
Where's the Boost?

**Comment by bbbales2** [comment]
@bgoodri On my hard drive ready to go. We decided to finish this pull req. off first and then do the Boost stuff.

**Comment by bob-carpenter** [comment]
???


> On Jun 21, 2018, at 3:34 PM, Ben Bales <notifications@github.com> wrote:
> 
> @bgoodri On my hard drive ready to go. We decided to finish this pull req. off first and then do the Boost stuff.

I'd rather not commit things to develop we're not happy to release.  If that's the intent, then this should be OK.   If it's just a placeholder that's going to get replaced before ever going live, then it should just be replaced straight off, I think---that is, you can stage the dev this way in a branch, but the intermediate thing shouldn't be committed if it's not release-ready.

**Comment by bbbales2** [comment]
My fault! I shoulda communicated more clearly. I’ll open a new pull req. with the Boost stuff. It’s based off develop but has the interface changes we discussed here.

Thanks for the discussion everyone! Thanks for getting this all rolling Marco! I’ll see you all in a new pull req. tomorrow!

**Pull request closed without merging**